include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,1723 @@
+    //  odi map does not contain a field domain for that object
+    // another: not singleton, odi map contains a field domain for that object
+
+    // Solution: meet non-singleton in odi map with singleton in base domain
+    //  Let s1 be the first state, s2 is second one.
+    //  Let Vs be the fields in that abstract object.
+    //  For Base in s1,
+    //  only_singleton = object in s2
+    //  object in s2 = top (empty)
+    //  s1 meet s2 :
+    //  only_singleton meet Base in s1,
+    //      Object (Vs) in s2 = top
+    // E.g.
+    //  s1: Base = { V_1 = 3, V_2 = 4 }, Object (V_1, V_2) = empty
+    //  s2: Base = { ... }, Object (V_1, V_2) = { V_1 <= V_2, V_2 <= 5 }
+    //  s1 meet s2:
+    //    b = Base in s1 meet { V_1 <= V_2, V_2 <= 5 }
+    //    Base = b meet s2 in Base
+    //    Object (V_1, V_2) = empty
+
+    // Other cases:
+    // one state: exists such an abstract object
+    // another: does not exist that abstract object
+    // Example1:
+    //  s1: Base = { V_1 = 3, V_2 = 6 }, Object (V_1, V_2) = empty
+    //  s2: Base = { ... }, Object (V_1, V_2) = empty
+    //  s1 meet s2 :
+    //    Base = Base^s1 meet Base^s2, Object (V_1, V_2) = empty
+
+    // Example2:
+    //  s1: Base = { ... }, Object (V_1, V_2) = { V_1 <= V_2, V_2 <= 5 }
+    //  s2: Base = { ... }, Object (V_1, V_2) = empty
+    //  s1 meet s2 :
+    //   Base = Base^s1 meet Base^s2,
+    //   Object (V_1, V_2) = { V_1 <= V_2, V_2 <= 5 }
+
+    // 1. meet / narrowing the parts of each state that do not require common
+    // renaming
+    obj_info_env_t out_obj_info_env(
+        is_meet ? left.m_obj_info_env & right.m_obj_info_env
+                : left.m_obj_info_env && right.m_obj_info_env);
+    // 2. meet / narrowing the subdomain for base and objects
+    // 2.1 meet / narrowing the base subdomain
+    base_abstract_domain_t out_base_dom =
+        base_abstract_domain_t(is_meet ? left.m_base_dom & right.m_base_dom
+                                       : left.m_base_dom && right.m_base_dom);
+
+    // 2.2 meet / narrowing the objects' domain over odi map
+    // In the real operation,
+    // the following operates on two odi maps containg the same keys
+    // even if it can performed without such assumption.
+    odi_map_t out_odi_map(is_meet ? left.m_odi_map & right.m_odi_map
+                                  : left.m_odi_map && right.m_odi_map);
+
+    // Handle case 3
+    for (auto it = left.m_obj_info_env.begin(); it != left.m_obj_info_env.end();
+         it++) {
+      obj_id_t rep = it->first;
+      const small_range &left_num_refs = it->second.refcount_val();
+      const small_range &right_num_refs =
+          right.m_obj_info_env.at(rep).refcount_val();
+      if (left_num_refs.is_one() &&
+          right_num_refs == small_range::oneOrMore()) {
+        meet_non_singleton_with_singleton(rep, out_base_dom, right.m_odi_map);
+        field_abstract_domain_t out_obj_dom; // top
```

Comment:
Here it seems that we are losing a lot of precision by assigning top. why? 
meet should be able to refine things and improve precision.